lib/helpers.go

```diff
@@ -0,0 +1,153 @@
+// called GetMaxPossibleVUs() and why there is no GetMaxUnplannedVUs() function.
+//
+// As an example, imagine that you have a scheduler with MaxPlannedVUs=20 and
+// MaxUnaplannedVUs=0, followed immediately after by another scheduler with
+// MaxPlannedVUs=10 and MaxUnaplannedVUs=10. If you MaxPlannedVUs for the whole
+// test is 20, and MaxUnaplannedVUs, but since those schedulers won't run
+// concurrently, MaxVUs for the whole test is not 30, rather it's 20, since 20
+// VUs will be sufficient to run the test.
+//
+// IMPORTANT 2: this has one very important exception. The manual execution
+// scheduler doesn't use the MaxUnplannedVUs (i.e. this function will return 0),
+// since their initialization and usage is directly controlled by the user and
+// is effectively bounded only by the resources of the machine k6 is running on.
+func GetMaxPossibleVUs(steps []ExecutionStep) (result uint64) {
+	for _, s := range steps {
+		stepMaxPossibleVUs := s.PlannedVUs + s.MaxUnplannedVUs
+		if stepMaxPossibleVUs > result {
+			result = stepMaxPossibleVUs
+		}
+	}
+	return result
+}
+
+// GetEndOffset returns the time offset of the last step of the execution plan,
+// and whether that step is a final one, i.e. whether the number of planned or
+// unplanned
+func GetEndOffset(steps []ExecutionStep) (lastStepOffset time.Duration, isFinal bool) {
+	if len(steps) == 0 {
+		return 0, true
+	}
+	lastStep := steps[len(steps)-1]
+	return lastStep.TimeOffset, (lastStep.PlannedVUs == 0 && lastStep.MaxUnplannedVUs == 0)
+}
+
+// A helper function for joining error messages into a single string
+func concatErrors(errors []error, separator string) string {
+	errStrings := make([]string, len(errors))
+	for i, e := range errors {
+		errStrings[i] = e.Error()
+	}
+	return strings.Join(errStrings, separator)
+}
+
+// StreamExecutionSteps launches a new goroutine and emits all execution steps
+// at their appropriate time offsets over the returned unbuffered channel. If
+// closeChanWhenDone is specified, it will close the channel after it sends the
+// last step. If it isn't, or if the context is cancelled, the internal
+// goroutine will be stopped, *but the channel will remain open*!
+//
+// As usual, steps in the supplied slice have to be sorted by their TimeOffset
+// values in an ascending order. Of course, multiple events can have the same
+// time offset (incl. 0).
+func StreamExecutionSteps(
+	ctx context.Context, startTime time.Time, steps []ExecutionStep, closeChanWhenDone bool,
+) <-chan ExecutionStep {
+
+	ch := make(chan ExecutionStep)
+	go func() {
+		for _, step := range steps {
+			offsetDiff := step.TimeOffset - time.Since(startTime)
+			if offsetDiff > 0 { // wait until time of event arrives
+				select {
+				case <-ctx.Done():
+					return // exit if context is cancelled
+				case <-time.After(offsetDiff): //TODO: reuse a timer?
+					// do nothing
+				}
+			}
+			select {
+			case <-ctx.Done():
```

Comment:
Not certain if we shouldn't have a previous select that is just 
```
select{
case <-ctx.Done():
return
default:
}
```
To be certain that if the context is `Done` k6 doesn't step the next step. Looking at the other code ... It might be unnecessary ... but still .. or at least adding a comment that this is not a problem because <something> will be helpful in the future when/if something breaks/changes and we are debugging.

Reply:
Yeah, I can add a comment, but since we're creating the output channel here and it's unbuffered, I don't think we need this extra `select`.

Reply:
added a comment